File: lib/libj2/Stream.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2020 all rights reserved


// get the forward declarations
#include "forward.h"
// external support
#include "externals.h"

// support for color
#include "ASCII.h"
#include "CSI.h"
// renderer support
#include "Renderer.h"
#include "Memo.h"
#include "Alert.h"

// get the device declaration
#include "Device.h"
// get the stream declaration
#include "Stream.h"
// get the manager of the global state
#include "Chronicler.h"


// metamethods
// destructor
pyre::journal::Stream::
~Stream()
{}


// interface
auto
pyre::journal::Stream::
memo(verbosity_type verbosity, const page_type & page, const metadata_type & meta) -> Stream &
{
    // get the verbosity level
    auto maxVerbosity = chronicler_t::verbosity();
    // if this message is chattier
    if (verbosity > maxVerbosity) {
        // do nothing
        return *this;
    }
    // otherwise, get the memo renderer to format the message
    auto content = _memo->render(_palette, page, meta);
    // inject it into my stream
    _stream << content;
    // all done
    return *this;
}


auto
pyre::journal::Stream::
alert(verbosity_type verbosity, const page_type & page, const metadata_type & meta) -> Stream &
{
    // get the verbosity level
    auto maxVerbosity = chronicler_t::verbosity();
    // if this message is chattier
    if (verbosity > maxVerbosity) {
        // do nothing
        return *this;
    }
    // otherwise, get the alert renderer to format the message
    auto content = _alert->render(_palette, page, meta);
    // inject it into my stream
    _stream << content;
    // all done
    return *this;
}


// end of file

File: pkg/j2/Stream.py
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2020 all rights reserved


# superclass
from .Device import Device


# write message to a stream
class Stream(Device):
    """
    Journal device that writes messages to a stream
    """


    # interface
    def close(self):
        """
        Close the associated stream
        """
        # delegate
        self.stream.close()
        # all done
        return self


    # metamethods
    def __init__(self, stream, name="stream", **kwds):
        # chain up
        super().__init__(name=name, **kwds)
        # save the stream
        self.stream = stream
        # all done
        return


    # implementation details
    def record(self, page):
        """
        Record a message
        """
        # assemble the content
        content = "\n".join(page)
        # if there is anything there
        if content:
            # inject it
            print(content, file=self.stream)
        # all done
        return


# end of file
